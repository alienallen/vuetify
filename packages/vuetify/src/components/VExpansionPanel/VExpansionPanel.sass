@import '../../styles/styles.sass'
@import './_mixins.sass'

// Theme
.v-expansion-panel.theme--light
  +expansion-panel-theme($material-light)
.v-expansion-panel.theme--dark
  +expansion-panel-theme($material-dark)

// Block
.v-expansion-panel
  border-radius: 4px
  display: flex
  flex-wrap: wrap
  justify-content: center
  list-style-type: none
  padding: 0
  width: 100%
  z-index: 1

  > *:first-child
    border-top-left-radius: inherit
    border-top-right-radius: inherit

  > *:last-child
    border-bottom-left-radius: inherit
    border-bottom-right-radius: inherit

// Element
.v-expansion-panel-item
  flex: 1 0 100%
  max-width: 100%
  position: relative
  transition: .2s map-get($transition, 'fast-in-slow-out')

  &:before
    border-radius: inherit
    content: ''
    position: absolute
    top: 0
    bottom: 0
    left: 0
    right: 0
    z-index: -1
    +elevationTransition()
    +elevation(1)

  &:not(:first-child):after
    content: ''
    left: 0
    top: 0
    height: 1px
    transition: .2s background-color map-get($transition, 'fast-out-slow-in')
    position: absolute
    right: 0

  &:not(.v-expansion-panel-item--has-mousedown)
    .v-expansion-panel-item__header:focus:before
      opacity: 0.12

  &--disabled
    .v-expansion-panel-item__header
      pointer-events: none

  &--active:not(.v-expansion-panel-item--disable-icon-rotate)
    > .v-expansion-panel-item__header
      .v-expansion-panel-item__header__icon .v-icon
        transform: rotate(-180deg)


.v-expansion-panel-item__header__icon
  display: inline-flex
  margin-top: -4px
  margin-left: auto
  margin-bottom: -4px
  user-select: none

  +rtl()
    margin-left: initial
    margin-right: auto

.v-expansion-panel-item__header
  align-items: center
  cursor: pointer
  display: flex
  font-size: 0.9375rem
  line-height: 1
  min-height: 48px
  outline: none
  padding: 16px
  position: relative

  &:before
    background-color: currentColor
    border-radius: inherit
    opacity: 0
    top: 0
    bottom: 0
    content: ''
    left: 0
    position: absolute
    right: 0
    transition: .3s opacity map-get($transition, 'fast-out-slow-in')

  > *:not(.v-expansion-panel-item__header__icon)
    flex: 1 1 auto

.v-expansion-panel-item__body
  > .v-card
    border-radius: 0
    +elevation(0)

// Modifier
.v-expansion-panel--focusable
  .v-expansion-panel-item--active
    .v-expansion-panel-item__header
      &:before
        opacity: 0.12

      &:focus
        &:before
          opacity: 0.2

  // .v-expansion-panel-item--active
  //   margin: 16px 0 0

.v-expansion-panel--accordion,
.v-expansion-panel--popout,
.v-expansion-panel--inset
  .v-expansion-panel-item--active
    margin: 16px 0

    + .v-expansion-panel-item--active
      margin-top: 0

    &:first-child
      margin: 0 0 16px

.v-expansion-panel--popout,
.v-expansion-panel--inset,
.v-expansion-panel--accordion
  .v-expansion-panel-item
    > *:first-child
      border-top-left-radius: inherit
      border-top-right-radius: inherit

    > *:last-child
      border-bottom-left-radius: inherit
      border-bottom-right-radius: inherit

  .v-expansion-panel-item__body
    > .v-card
      border-radius: inherit

  .v-expansion-panel-item--active
    border-radius: 4px

    &:before
      +elevation(2)

    &:after,
    + .v-expansion-panel-item:after
      opacity: 0

.v-expansion-panel--popout
  .v-expansion-panel-item
    max-width: calc(100% - 32px)

.v-expansion-panel--inset
  .v-expansion-panel-item
    max-width: 100%

.v-expansion-panel--popout
  .v-expansion-panel-item--active
    max-width: calc(100% + 16px)

.v-expansion-panel--inset
  .v-expansion-panel-item--active
    max-width: calc(100% - 32px)
