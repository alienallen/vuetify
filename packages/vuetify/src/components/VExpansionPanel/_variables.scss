$expansion-panel-border-radius: 4px !default;
$expansion-panel-item-active-margin: 16px !default;
$expansion-panel-item-header-font-size: 0.9375rem !default;
$expansion-panel-item-header-min-height: 48px !default;
$expansion-panel-item-header-padding: 16px !default;
$expansion-panel-popout-max-width: calc(100% - #{$expansion-panel-item-active-margin * 2});
$expansion-panel-popout-active-max-width: calc(100% + #{$expansion-panel-item-active-margin});
$expansion-panel-inset-max-width: 100%;
$expansion-panel-inset-active-max-width: calc(100% - #{$expansion-panel-item-active-margin * 2});
