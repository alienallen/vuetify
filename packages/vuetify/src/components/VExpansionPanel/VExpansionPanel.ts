// Styles
import './VExpansionPanel.sass'

// Extensions
import { BaseItemGroup } from '../VItemGroup/VItemGroup'

// Utilities
import { breaking } from '../../util/console'

// Types
import { VNode } from 'vue'

/* @vue/component */
export default BaseItemGroup.extend({
  name: 'v-expansion-panel',

  provide (): object {
    return {
      expansionPanel: this
    }
  },

  props: {
    accordion: Boolean,
    disabled: Boolean,
    readonly: Boolean,
    expand: Boolean,
    focusable: Boolean,
    inset: Boolean,
    popout: Boolean
  },

  computed: {
    classes (): object {
      return {
        ...BaseItemGroup.options.computed.classes.call(this),
        'v-expansion-panel': true,
        'v-expansion-panel--accordion': this.accordion,
        'v-expansion-panel--focusable': this.focusable,
        'v-expansion-panel--popout': this.popout,
        'v-expansion-panel--inset': this.inset
      }
    }
  },

  created () {
    /* istanbul ignore next */
    if (this.expand) breaking('expand', 'multiple')

    /* istanbul ignore next */
    if (
      Array.isArray(this.value) &&
      this.value.length > 0 &&
      typeof this.value[0] === 'boolean'
    ) {
      breaking('<v-expansion-panel :value="[true, false]">', '<v-expansion-panel :value="[0]">')
    }
  }
})
