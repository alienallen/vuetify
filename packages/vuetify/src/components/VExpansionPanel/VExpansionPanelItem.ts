// Components
import VExpansionPanel from './VExpansionPanel'
import { VExpandTransition } from '../transitions'
import VIcon from '../VIcon'

// Mixins
import { factory as GroupableFactory } from '../../mixins/groupable'
import Bootable from '../../mixins/bootable'
import Rippleable from '../../mixins/rippleable'

// Utilities
import { keyCodes } from '../../util/helpers'
import mixins, { ExtractVue } from '../../util/mixins'

// Types
import { VNode } from 'vue'

const baseMixins = mixins(
  Bootable,
  GroupableFactory('expansionPanel'),
  Rippleable
)

interface options extends ExtractVue<typeof baseMixins> {
  $refs: {
    header: HTMLElement
  }
  expansionPanel: InstanceType<typeof VExpansionPanel>
}

/* @vue/component */
export default baseMixins.extend<options>().extend({
  name: 'v-expansion-panel-item',

  props: {
    disabled: Boolean,
    readonly: Boolean,
    expandIcon: {
      type: String,
      default: '$vuetify.icons.expand'
    },
    hideActions: Boolean,
    disableIconRotate: Boolean,
    ripple: {
      type: [Boolean, Object],
      default: false
    }
  },

  data: () => ({
    height: 'auto',
    hasMousedown: false
  }),

  computed: {
    classes (): object {
      return {
        'v-expansion-panel-item--active': this.isActive,
        'v-expansion-panel-item--disabled': this.isDisabled,
        'v-expansion-panel-item--has-mousedown': this.hasMousedown,
        'v-expansion-panel-item--disable-item-rotate': this.disableIconRotate,
        ...this.groupClasses
      }
    },
    isDisabled (): boolean {
      return this.expansionPanel.disabled || this.disabled
    },
    isReadonly (): boolean {
      return this.expansionPanel.readonly || this.readonly
    }
  },

  methods: {
    onKeydown (e: KeyboardEvent) {
      // Ensure element is the activeElement
      if (
        e.keyCode === keyCodes.enter &&
        this.$refs.header === document.activeElement
      ) this.click(e)
    },
    click (e: MouseEvent | KeyboardEvent) {
      if (e.detail) this.$refs.header.blur()

      this.$emit('click', e)

      this.isReadonly || this.toggle()
    },
    genBody () {
      return this.$createElement('div', {
        class: 'v-expansion-panel-item__body',
        directives: [{
          name: 'show',
          value: this.isActive
        }]
      }, this.showLazyContent(this.$slots.default))
    },
    genHeader () {
      const children = [...(this.$slots.header || [])]

      if (!this.hideActions) children.push(this.genIcon())

      return this.$createElement('div', {
        staticClass: 'v-expansion-panel-item__header',
        attrs: {
          tabindex: this.isReadonly || this.isDisabled ? null : 0
        },
        directives: [{
          name: 'ripple',
          value: this.ripple
        }],
        on: {
          click: this.click,
          keydown: this.onKeydown,
          mousedown: this.onMousedown,
          mouseup: this.onMouseup
        },
        ref: 'header'
      }, children)
    },
    genIcon () {
      const icon = this.$slots.actions ||
        [this.$createElement(VIcon, this.expandIcon)]

      return this.$createElement('transition', {
        attrs: { name: 'fade-transition' }
      }, [
        this.$createElement('div', {
          staticClass: 'v-expansion-panel-item__header__icon',
          directives: [{
            name: 'show',
            value: !this.isDisabled
          }]
        }, icon)
      ])
    },
    onMousedown (e: Event) {
      this.hasMousedown = true
      this.$emit('mousedown', e)
    },
    onMouseup (e: Event) {
      this.hasMousedown = false
      this.$emit('mouseup', e)
    },
    toggle () {
      this.isBooted = true

      this.$nextTick(() => this.$emit('change'))
    }
  },

  render (h): VNode {
    return h('li', {
      staticClass: 'v-expansion-panel-item',
      class: this.classes,
      attrs: {
        'aria-expanded': String(!!this.isActive)
      }
    }, [
      this.$slots.header && this.genHeader(),
      h(VExpandTransition, [this.genBody()])
    ])
  }
})
