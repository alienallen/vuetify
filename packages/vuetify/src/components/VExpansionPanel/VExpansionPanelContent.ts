import VExpansionPanelItem from './VExpansionPanelItem'
import { deprecate } from '../../util/console'

/* @vue/component */
export default VExpansionPanelItem.extend({
  created () {
    deprecate('<v-expansion-panel-content>', '<v-expansion-panel-item>')
  }
})
