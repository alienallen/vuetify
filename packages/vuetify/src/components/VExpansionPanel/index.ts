import VExpansionPanel from './VExpansionPanel'
import VExpansionPanelItem from './VExpansionPanelItem'
import VExpansionPanelContent from './VExpansionPanelContent'

export { VExpansionPanel, VExpansionPanelContent, VExpansionPanelItem }

export default {
  $_vuetify_subcomponents: {
    VExpansionPanel,
    VExpansionPanelContent,
    VExpansionPanelItem
  }
}
