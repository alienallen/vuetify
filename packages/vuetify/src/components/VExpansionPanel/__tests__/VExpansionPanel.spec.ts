// Components
import VExpansionPanel from '../VExpansionPanel'

// Utilities
import {
  mount,
  Wrapper
} from '@vue/test-utils'

describe('VExpansionPanel.ts', () => {
  type Instance = InstanceType<typeof VExpansionPanel>
  let mountFunction: (options?: object) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(VExpansionPanel, {
        ...options
      })
    }
  })

  it('should work', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
