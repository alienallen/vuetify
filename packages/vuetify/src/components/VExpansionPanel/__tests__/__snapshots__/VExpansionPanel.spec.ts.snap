// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VExpansionPanel.ts should work 1`] = `

<div class="v-item-group theme--light v-expansion-panel">
</div>

`;
